### DISCUSSION TITLE
Fitting Discontinuous Expressions

### DISCUSSION THREAD

**Discussion by janzenchoi** [MilesCranmer/PySR#1007] [Q&A] [answered]
Hello,

I am using `TemplateExpressionSpec` to fit two expressions, `z0` and `z1`. The `z0` expression is undefined/complex after `x0 > z1`. The template returns the sum of the relative errors, which the loss function interprets as the loss. The following is a snippet of my code to fit the discontinuous expression.

```
expression_spec = TemplateExpressionSpec(
    expressions    = ["f0", "fA", "fn", "fM", "fphi", "fchi"],
    variable_names = ["x0", "x1", "x2", "y0", "y1"],
    combine        = """
        A = 10^fA(x2); n = abs(fn(x2)); M = 10^fM(x2); phi = abs(fphi(x2)); chi = abs(fchi(x2));
        z0 = A*x1^n * ((1-(phi+1)*M*x1^chi*x0)^((phi+1-n)/(phi+1))-1) / (M*x1^chi*(n-phi-1)) + f0(x0,x1,x2);
        z1 = 1/((phi+1)*M*x1^chi);
        ((y0-z0)/y0)^2 + ((y1-z1)/y1)^2
    """
)

regressor = PySRRegressor(
    expression_spec  = expression_spec,
    populations      = 32,
    population_size  = 32,
    maxsize          = 16,
    niterations      = 128,
    precision        = 64,
    verbosity        = 0,
    binary_operators = ["+", "*", "/"],
    unary_operators  = ["log"],
    elementwise_loss = "take_first(p, t, w) = p",
)
```

In fitting the expressions, the loss function may try to evaluate the relative error where the `z0` is undefined (i.e., `x0 > z1`).

Is there any way to have the loss function only calculate the loss for defined / real `z0` values (i.e., when `x0 < z1 `)? 

Thanks again for the help.

**Comment by MilesCranmer** [comment]
Not sure if I follow butiIt should automatically do this already. The NaNs will propagate and the expression will receive an infinite loss. Therefore the genetic algorithm will avoid expression that give such terms. 

But yes, it will consider all points in the dataset. If _any_ point is NaN, it will be thrown out.

**Reply by janzenchoi** [reply]
For my problem, a solution can be good even if there are NaNs when `x0 > z1`. Is there a way for the solution to not be thrown out if there are NaNs? Instead, the loss function would only calculate the error for `x0 < z1`.

**Reply by MilesCranmer** [reply, marked as answer]
what happens if all x0 > z1? Or if only one point is safe? I guess you’d have to set some threshold.

but you can get the underlying arrays with `.x` and the validity with `.valid`. Just be sure to propagate the validity and return a `ValidVector(x, valid)` from the code block.

**Reply by janzenchoi** [reply]
I included the relative error for `z1` in the loss function, so I'm thinking that having only one safe point wouldn't be a big issue.

I'm not very familiar with Julia. Where would I access the `.x` and `.valid` values? And how would one do such propagation?

**Reply by MilesCranmer** [reply]
All of the variables in the `combine` code are ValidVector types. They each have `.valid` and `.x`
